### PR TITLE
Add a behavior for using arrow keys and other types of keys to move focus

### LINKED ISSUE #1121: Fix `isTabbable` logic
**Describe the bug**
With the current implementation, `isTabbable` returns true for some elements that are not tabbable.

**To Reproduce**
1. Render this element: `<input type="hidden" tabindex="0" />`
2. Call `isTabbable` on that element

**Expected behavior**
False

**Additional context**
To be resolved in `arrowKeyFocus` behavior PR.

### LINKED ISSUE #1134: Behavior: Arrow Key Focus
## Description
We need a behavior that sets up focus management with arrow keys (and potentially other keys). According to [ARIA Best Practices](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_generalnav), there are many cases where focus should be moved using arrow keys, but the DOM only has native support for managing focus using the Tab key.

Furthermore, we should support the [`aria-activedescendant` pattern](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_focus_activedescendant), which will be used in "filter list" UI and potentially other situations.

Due to the complexity of this behavior, it should be written in "vanilla" JS and wrapped by a hook.

### Specification & Usage
See the [proposed documentation](https://primer-components-f5ts87di8-primer.vercel.app/components/focusZone) for more details.

### PR DESCRIPTION
**Pull request by T-Hugs**
This PR introduces the `focusZone` behavior along with its companion hook, `useFocusZone`. Focus zones are used to allow keyboard navigation within a container using keys other than Tab, such as the arrow keys, home and end, and HJKL.

For more details about `focusZone`, see the [rendered documentation](https://primer-components-d5qlwhk4x-primer.vercel.app/components/focusZone).

Closes #1134
Closes #1121

### Screenshots
This PR includes 7 Storybook stories. Here is a screenshot of one of them.
image

### Merge checklist
- [x] Added or updated TypeScript definitions (`index.d.ts`) if necessary
- [x] Added/updated tests
- [x] Added/updated documentation
- [x] Tested in Chrome
- [x] Tested in Firefox
- [x] Tested in Safari
- [ ] Tested in Edge

### CODE REVIEW

**Review by dgreif — CHANGES_REQUESTED** [review]
_Very_ thorough implementation!  Most of my suggestions are cosmetic, the guts of focusTrap seem really solid.  Nice work! ➡️  🏅

File: src/behaviors/focusZone.ts [hunk 1]
```diff
@@ -226,6 +226,14 @@ function getDirection(keyboardEvent: KeyboardEvent) {
   if (keyboardEvent.key === 'Tab' && keyboardEvent.shiftKey) {
     return 'previous'
   }
+  const isMac = isMacOS()
+  if ((isMac && keyboardEvent.metaKey) || (!isMac && keyboardEvent.ctrlKey)) {
```
[the remark this answers is not in the document]
Comment (T-Hugs) [reply src/behaviors/focusZone.ts:230]: Going to keep as-is for now. I think it's slightly more readable as-is.

File: src/behaviors/focusZone.ts [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+      }
+      element.setAttribute('tabindex', '-1')
+    }
+  }
+
+  function endFocusManagement(...elements: HTMLElement[]) {
+    for (const element of elements) {
+      const tabbableElementIndex = focusableElements.indexOf(element)
+      if (tabbableElementIndex >= 0) {
+        focusableElements.splice(tabbableElementIndex, 1)
```
[the remark this answers is not in the document]
Comment (T-Hugs) [reply src/behaviors/focusZone.ts:401]: 👍 You can probably tell this was before I added the `isFocusable`/`isTabbable` distinction :)

File: src/behaviors/focusZone.ts [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+
+const subscriptions: ((activeElement: HTMLElement) => void)[] = []
+function notifyActiveElement(element: HTMLElement) {
+  for (const subscription of subscriptions) {
+    subscription(element)
+  }
+}
+
+function subscribeToActiveElementChanges(callback: (activeElement: HTMLElement) => void) {
+  subscriptions.push(callback)
```
[the remark this answers is not in the document]
Comment (T-Hugs) [reply src/behaviors/focusZone.ts:324]: My brain hurts.

[the remark this answers is not in the document]
Comment (T-Hugs) [reply src/behaviors/focusZone.ts:318]: Fixed.

**Comment by T-Hugs** [comment]
@dgreif Thanks for the review! Everything addressed. Let me know if you see anything else.

**Review by dgreif — APPROVED** [review]

**Pull request merged**